csharp/src/Apache.Arrow/Schema.cs

```diff
@@ -44,80 +43,74 @@ public partial class Schema
             IEnumerable<Field> fields,
             IEnumerable<KeyValuePair<string, string>> metadata)
         {
-            if (fields == null)
+            if (fields is null)
             {
                 throw new ArgumentNullException(nameof(fields));
             }
 
-            _fields = fields.ToList();
-
-            _fieldsDictionary = fields.ToDictionary(field => field.Name, field => field);
+            _fieldsList = fields.ToList();
+            FieldsLookup = _fieldsList.ToLookup(f => f.Name);
+            _fieldsDictionary = FieldsLookup.ToDictionary(g => g.Key, g => g.First());
 
             Metadata = metadata?.ToDictionary(kv => kv.Key, kv => kv.Value);
         }
 
-        internal Schema(List<Field> fields, IReadOnlyDictionary<string, string> metadata, bool copyCollections)
+        internal Schema(List<Field> fieldsList, IReadOnlyDictionary<string, string> metadata, bool copyCollections)
         {
-            Debug.Assert(fields != null);
+            Debug.Assert(fieldsList != null);
             Debug.Assert(copyCollections == false, "This internal constructor is to not copy the collections.");
 
-            _fields = fields;
-
-            _fieldsDictionary = fields.ToDictionary(field => field.Name, field => field);
+            _fieldsList = fieldsList;
+            FieldsLookup = _fieldsList.ToLookup(f => f.Name);
+            _fieldsDictionary = FieldsLookup.ToDictionary(g => g.Key, g => g.First());
 
             Metadata = metadata;
         }
 
-        public Field GetFieldByIndex(int i)
-        {
-            return _fields[i];
-        }
+        public Field GetFieldByIndex(int i) => _fieldsList[i];
 
-        public Field GetFieldByName(string name) =>
-            Fields.TryGetValue(name, out Field field) ? field : null;
+        public Field GetFieldByName(string name) => FieldsLookup[name].First();
```

Comment:
```suggestion
        public Field GetFieldByName(string name) => FieldsLookup[name].FirstOrDefault();
```

Reply:
Can you add a test here to ensure `null` is returned when a field doesn't exist?

Reply:
DOne